ompi/mca/pml/cm/pml_cm_sendreq.h

```diff
@@ -382,28 +382,30 @@ do {                                                                    \
-                                  sendreq->req_send.req_base.req_comm,           \
-                                  sendreq->req_peer,                             \
-                                  sendreq->req_tag,                              \
-                                  &sendreq->req_send.req_base.req_convertor,     \
-                                  sendreq->req_send.req_send_mode,               \
-                                  sendreq->req_blocking,                         \
-                                  &sendreq->req_mtl));                           \
-        if(OMPI_SUCCESS == ret &&                                                \
-           sendreq->req_send.req_send_mode == MCA_PML_BASE_SEND_BUFFERED) {      \
-            sendreq->req_send.req_base.req_ompi.req_status.MPI_ERROR = 0;        \
-            ompi_request_complete(&(sendreq)->req_send.req_base.req_ompi, true); \
-        }                                                                        \
-    }                                                                            \
+#define MCA_PML_CM_HVY_SEND_REQUEST_START(sendreq, ret)                              \
+do {                                                                                 \
+    ret = OMPI_SUCCESS;                                                              \
+    MCA_PML_CM_SEND_REQUEST_START_SETUP(&(sendreq)->req_send);                       \
+    if (sendreq->req_send.req_send_mode == MCA_PML_BASE_SEND_BUFFERED) {             \
+        MCA_PML_CM_HVY_SEND_REQUEST_BSEND_ALLOC(sendreq, ret);                       \
+    }                                                                                \
+    if (OMPI_SUCCESS == ret) {                                                       \
+        ret = OMPI_MTL_CALL(isend(ompi_mtl,                                          \
+                                  sendreq->req_send.req_base.req_comm,               \
+                                  sendreq->req_peer,                                 \
+                                  sendreq->req_tag,                                  \
+                                  &sendreq->req_send.req_base.req_convertor,         \
+                                  sendreq->req_send.req_send_mode,                   \
+                                  sendreq->req_blocking,                             \
+                                  &sendreq->req_mtl));                               \
+        if(OMPI_SUCCESS == ret &&                                                    \
+           sendreq->req_send.req_send_mode == MCA_PML_BASE_SEND_BUFFERED) {          \
+            sendreq->req_send.req_base.req_ompi.req_status.MPI_ERROR = 0;            \
+            if(!REQUEST_COMPLETE(&sendreq->req_send.req_base.req_ompi)) {            \
+                ompi_request_complete(&(sendreq)->req_send.req_base.req_ompi, true); \
```

Comment:
This needs a comment (something like `/* the isend call may have marked completion */`).